src/expr/transform/reduction.rs

```diff
@@ -359,8 +359,22 @@ impl FoldConstants {
         // will be consolidated. We have to make a separate check for constant
         // nodes here, since the match arm above might install new constant
         // nodes.
-        if let RelationExpr::Constant { rows, .. } = relation {
+        if let RelationExpr::Constant { rows, typ } = relation {
+            // Reduce down to canonical representation.
             differential_dataflow::consolidation::consolidate(rows);
+            rows.retain(|(_row, count)| count != &0);
```

Comment:
I’m still skeptical that this world where `differential_dataflow::consolidate` needs to leave zero-diffed records around for performance will ever come to pass. But since you seem to think it might, should we extract our own `consolidate_more` function with these two lines,  and stick a comment on it? I can totally see future me going “oh right here’s how to consolidate diffs,” copy paste the first line, and leave behind the second, because it’s not obviously related (to me, as currently written).

Reply:
I think if there are meant to be invariants that are different from differential dataflow's, for example that materialize panics on zero diffs, and we have a hard time maintaining them, we should probably either test things or write a bunch of helpers to avoid those mistakes in the future.

Reply:
We should probably also document what those invariants are. For example, differential dataflow's `consolidate` will put rows in sorted order, which is important for it, but may or may not be important for us.